File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.cs
public class Solution {
    public int[] Intersect(int[] nums1, int[] nums2) {
        Dictionary<int, int> counts = new Dictionary<int, int>();
        
        // Count the occurrences of each element in nums1
        foreach (int num in nums1)
        {
            if (counts.ContainsKey(num))
            {
                counts[num]++;
            }
            else
            {
                counts[num] = 1;
            }
        }
        
        List<int> result = new List<int>();
        
        // Traverse nums2 and find common elements
        foreach (int num in nums2)
        {
            if (counts.ContainsKey(num) && counts[num] > 0)
            {
                result.Add(num);
                counts[num]--;
            }
        }
        
        return result.ToArray();
    }
}

File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int,int>ump;
        for(int i=0; i<nums1.size(); i++)
        {
            ump[nums1[i]]++;
        }
        
        vector<int>result;
        for(auto it:nums2)
        {
            if(ump[it]>0)
            {
                result.push_back(it);
                ump[it]--;
            }
        }
        return result;
    }
};